File: CBLClient/Apps/CBLTestServer-Android/app/src/main/java/com/couchbase/CouchbaseLiteServ/server/RequestHandler/FunctionRequestHandler.java
package com.couchbase.CouchbaseLiteServ.server.RequestHandler;


import com.couchbase.CouchbaseLiteServ.server.Args;
import com.couchbase.lite.Expression;
import com.couchbase.lite.Function;

public class FunctionRequestHandler {

    public Expression avg(Args args){
        Expression expression = args.get("expression");
        return Function.avg(expression);
    }

    public Expression count(Args args){
        Expression expression = args.get("expression");
        return Function.count(expression);
    }

    public Expression min(Args args){
        Expression expression = args.get("expression");
        return Function.min(expression);
    }

    public Expression max(Args args){
        Expression expression = args.get("expression");
        return Function.max(expression);
    }

    public Expression sum(Args args){
        Expression expression = args.get("expression");
        return Function.sum(expression);
    }

    public Expression abs(Args args){
        Expression expression = args.get("expression");
        return Function.abs(expression);
    }

    public Expression acos(Args args){
        Expression expression = args.get("expression");
        return Function.acos(expression);
    }

    public Expression asin(Args args){
        Expression expression = args.get("expression");
        return Function.asin(expression);
    }

    public Expression atan(Args args){
        Expression expression = args.get("expression");
        return Function.atan(expression);
    }

    public Expression atan2(Args args){
        Expression x = args.get("x");
        Expression y = args.get("y");
        return Function.atan2(x, y);
    }

    public Expression ceil(Args args){
        Expression expression = args.get("expression");
        return Function.ceil(expression);
    }

    public Expression cos(Args args){
        Expression expression = args.get("expression");
        return Function.cos(expression);
    }

    public Expression degrees(Args args){
        Expression expression = args.get("expression");
        return Function.degrees(expression);
    }

    public Expression e(){
        return Function.e();
    }

    public Expression exp(Args args){
        Expression expression = args.get("expression");
        return Function.exp(expression);
    }

    public Expression floor(Args args){
        Expression expression = args.get("expression");
        return Function.floor(expression);
    }

    public Expression ln(Args args){
        Expression expression = args.get("expression");
        return Function.ln(expression);
    }

    public Expression log(Args args){
        Expression expression = args.get("expression");
        return Function.log(expression);
    }

    public Expression pi(Args args){
        return Function.pi();
    }

    public Expression power(Args args){
        Expression base = args.get("base");
        Expression exponent = args.get("exponent");
        return Function.power(base, exponent);
    }

    public Expression radians(Args args){
        Expression expression = args.get("expression");
        return Function.radians(expression);
    }

    public Expression round(Args args){
        Expression expression = args.get("expression");
        return Function.round(expression);
    }

    public Expression roundDigits(Args args){
        Expression expression = args.get("expression");
        Expression digits = args.get("digits");
        return Function.round(expression, digits);
    }

    public Expression sign(Args args){
        Expression expression = args.get("expression");
        return Function.sign(expression);
    }

    public Expression sin(Args args){
        Expression expression = args.get("expression");
        return Function.sin(expression);
    }

    public Expression sqrt(Args args){
        Expression expression = args.get("expression");
        return Function.sqrt(expression);
    }

    public Expression tan(Args args){
        Expression expression = args.get("expression");
        return Function.tan(expression);
    }

    public Expression trunc(Args args){
        Expression expression = args.get("expression");
        return Function.trunc(expression);
    }

    public Expression truncDigits(Args args){
        Expression expression = args.get("expression");
        Expression digits = args.get("digits");
        return Function.trunc(expression,digits);
    }

    public Expression contains(Args args){
        Expression expression = args.get("expression");
        Expression substring = args.get("substring");
        return Function.contains(expression, substring);
    }

    public Expression length(Args args){
        Expression expression = args.get("expression");
        return Function.length(expression);
    }

    public Expression lower(Args args){
        Expression expression = args.get("expression");
        return Function.lower(expression);
    }

    public Expression ltrim(Args args){
        Expression expression = args.get("expression");
        return Function.ltrim(expression);
    }

    public Expression rtrim(Args args){
        Expression expression = args.get("expression");
        return Function.rtrim(expression);
    }

    public Expression trim(Args args){
        Expression expression = args.get("expression");
        return Function.trim(expression);
    }

    public Expression upper(Args args){
        Expression expression = args.get("expression");
        return Function.upper(expression);
    }

}

File: CBLClient/Apps/CBLTestServer-iOS/CBLTestServer-iOS/Server/FunctionRequestHandler.swift
//
//  FunctionRequestHandler.swift
//  CBLTestServer-iOS
//
//  Created by Raghu Sarangapani on 1/16/18.
//  Copyright © 2018 Raghu Sarangapani. All rights reserved.
//

import Foundation
import CouchbaseLiteSwift

public class FunctionRequestHandler {
    public static let VOID: String? = nil
    fileprivate var _pushPullReplListener:NSObjectProtocol?
    
    public func handleRequest(method: String, args: Args) throws -> Any? {
        switch method {
        //////////////
        // Function //
        //////////////
        case "function_avg":
            let expression: Expression = args.get(name: "expression")!
            return Function.avg(expression as! ExpressionProtocol)
            
        case "function_count":
            let expression: Expression = args.get(name: "expression")!
            return Function.count(expression as! ExpressionProtocol)
            
        case "function_min":
            let expression: Expression = args.get(name: "expression")!
            return Function.min(expression as! ExpressionProtocol)
            
        case "function_max":
            let expression: Expression = args.get(name: "expression")!
            return Function.max(expression as! ExpressionProtocol)
            
        case "function_sum":
            let expression: Expression = args.get(name: "expression")!
            return Function.sum(expression as! ExpressionProtocol)
            
        case "function_abs":
            let expression: Expression = args.get(name: "expression")!
            return Function.abs(expression as! ExpressionProtocol)
            
        case "function_acos":
            let expression: Expression = args.get(name: "expression")!
            return Function.acos(expression as! ExpressionProtocol)
            
        case "function_asin":
            let expression: Expression = args.get(name: "expression")!
            return Function.asin(expression as! ExpressionProtocol)
            
        case "function_atan":
            let expression: Expression = args.get(name: "expression")!
            return Function.atan(expression as! ExpressionProtocol)
            
        case "function_atan2":
            let x: Float = args.get(name: "x")!
            let y: Float = args.get(name: "y")!
            
            return Function.atan2(x:Expression.float(x), y:Expression.float(y))
            
        case "function_ceil":
            let expression: Expression = args.get(name: "expression")!
            return Function.ceil(expression as! ExpressionProtocol)
            
        case "function_cos":
            let expression: Expression = args.get(name: "expression")!
            return Function.cos(expression as! ExpressionProtocol)
            
        case "function_degrees":
            let expression: Expression = args.get(name: "expression")!
            return Function.degrees(expression as! ExpressionProtocol)
            
        case "function_e":
            return Function.e()
            
        case "function_exp":
            let expression: Expression = args.get(name: "expression")!
            return Function.exp(expression as! ExpressionProtocol)
            
        case "function_floor":
            let expression: Expression = args.get(name: "expression")!
            return Function.floor(expression as! ExpressionProtocol)
            
        case "function_ln":
            let expression: Expression = args.get(name: "expression")!
            return Function.ln(expression as! ExpressionProtocol)
            
        case "function_log":
            let expression: Expression = args.get(name: "expression")!
            return Function.log(expression as! ExpressionProtocol)
            
        case "function_pi":
            return Function.pi()
            
        case "function_power":
            let base: Int = args.get(name: "base")!
            let exponent: Int = args.get(name: "exponent")!
            
            return Function.power(base: Expression.int(base), exponent:Expression.int(exponent))
            
        case "function_radians":
            let expression: Expression = args.get(name: "expression")!
            return Function.radians(expression as! ExpressionProtocol)
            
        case "function_round":
            let expression: Expression = args.get(name: "expression")!
            return Function.round(expression as! ExpressionProtocol)
            
        case "function_roundDigits":
            let expression: Expression = args.get(name: "expression")!
            let digits: Int = args.get(name: "digits")!
            
            return Function.round(expression as! ExpressionProtocol, digits: Expression.int(digits))
            
        case "function_sign":
            let expression: Expression = args.get(name: "expression")!
            return Function.sign(expression as! ExpressionProtocol)
            
        case "function_sin":
            let expression: Expression = args.get(name: "expression")!
            return Function.sin(expression as! ExpressionProtocol)
            
        case "function_sqrt":
            let expression: Expression = args.get(name: "expression")!
            return Function.sqrt(expression as! ExpressionProtocol)
            
        case "function_tan":
            let expression: Expression = args.get(name: "expression")!
            return Function.tan(expression as! ExpressionProtocol)
            
        case "function_trunc":
            let expression: Expression = args.get(name: "expression")!
            return Function.trunc(expression as! ExpressionProtocol)
            
        case "function_truncDigits":
            let expression: Expression = args.get(name: "expression")!
            let digits: Int = args.get(name: "digits")!
            
            return Function.trunc(expression as! ExpressionProtocol, digits: Expression.int(digits))
            
        case "function_contains":
            let expression: Expression = args.get(name: "expression")!
            let substring: Expression = args.get(name: "substring")!
            return Function.contains(expression as! ExpressionProtocol, substring: substring as! ExpressionProtocol)
            
        case "function_length":
            let expression: Expression = args.get(name: "expression")!
            return Function.length(expression as! ExpressionProtocol)
            
        case "function_lower":
            let expression: Expression = args.get(name: "expression")!
            return Function.lower(expression as! ExpressionProtocol)
            
        case "function_ltrim":
            let expression: Expression = args.get(name: "expression")!
            return Function.ltrim(expression as! ExpressionProtocol)
            
        case "function_rtrim":
            let expression: Expression = args.get(name: "expression")!
            return Function.rtrim(expression as! ExpressionProtocol)
            
        case "function_trim":
            let expression: Expression = args.get(name: "expression")!
            return Function.trim(expression as! ExpressionProtocol)
            
        case "function_upper":
            let expression: Expression = args.get(name: "expression")!
            return Function.upper(expression as! ExpressionProtocol)
            
        case "function_rank":
            let indexName: String = args.get(name: "expression")!
            return FullTextFunction.rank(indexName)

        default:
            throw RequestHandlerError.MethodNotFound(method)
        }
        return FunctionRequestHandler.VOID
    }
}
